File: ui/sentry/scrubbers.js
const MEDIA_EXT = '(mkv|mp4|avi|webm|m4v|mov|mp3|flac|srt|vtt|ass)'
const HOME_PATH_RE = /\/(Users|home)\/[^/"'`<>]+?\//g
const MEDIA_FILE_RE = new RegExp(`[\\w.\\-]+\\.${MEDIA_EXT}`, 'gi')
const NUMERIC_ID_RE = /\/\d+(?=\/|$|\?)/g
const UUID_RE = /\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}(?=\/|$|\?)/gi
const SEARCH_TERM_RE = /(Failed to fetch (?:TVMaze|IMDb)[^:]*:\s*)(.+)/g

export function scrubString(input) {
  if (typeof input !== 'string') return input
  return input
    .replace(HOME_PATH_RE, '~/')
    .replace(MEDIA_FILE_RE, (match) => {
      const dot = match.lastIndexOf('.')
      return `*${match.slice(dot).toLowerCase()}`
    })
    .replace(SEARCH_TERM_RE, '$1<redacted>')
}

export function scrubUrl(input) {
  if (typeof input !== 'string') return input
  const [base] = input.split('?')
  return base
    .replace(UUID_RE, '/:id')
    .replace(NUMERIC_ID_RE, '/:id')
}

export function beforeSend(event) {
  if (!event) return event
  if (event.message) event.message = scrubString(event.message)
  if (event.request) {
    if (event.request.url) event.request.url = scrubUrl(event.request.url)
    if (event.request.query_string) event.request.query_string = ''
  }
  const values = event.exception?.values
  if (Array.isArray(values)) {
    for (const v of values) {
      if (v.value) v.value = scrubString(v.value)
      const frames = v.stacktrace?.frames
      if (Array.isArray(frames)) {
        for (const f of frames) {
          if (f.filename) f.filename = scrubString(f.filename)
          if (f.abs_path) f.abs_path = scrubString(f.abs_path)
        }
      }
    }
  }
  if (event.user?.username) delete event.user.username
  return event
}

export function beforeBreadcrumb(crumb) {
  if (!crumb) return crumb
  if (crumb.message) crumb.message = scrubString(scrubUrl(crumb.message))
  if (crumb.data) {
    if (crumb.data.url) crumb.data.url = scrubUrl(crumb.data.url)
    if (crumb.data.to) crumb.data.to = scrubUrl(crumb.data.to)
    if (crumb.data.from) crumb.data.from = scrubUrl(crumb.data.from)
  }
  return crumb
}

File: server/lib/sentry/scrubbers.rb
# Sentry PII scrubbers — Ruby port of ui/sentry/scrubbers.js.
# See docs/superpowers/specs/2026-04-22-sentry-integration-design.md
# for the "aggressive scrubbing" rationale.
module Sentry
  module Scrubbers
    MEDIA_EXT = "(mkv|mp4|avi|webm|m4v|mov|mp3|flac|srt|vtt|ass)".freeze
    HOME_PATH_RE = %r{/(Users|home)/[^/"'`<>]+?/}.freeze
    MEDIA_FILE_RE = /[\w.\-]+\.#{MEDIA_EXT}/i.freeze
    NUMERIC_ID_RE = %r{/\d+(?=/|\z|\?)}.freeze
    UUID_RE = %r{/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}(?=/|\z|\?)}i.freeze
    SEARCH_TERM_RE = /(Failed to fetch (?:TVMaze|IMDb)[^:]*:\s*)(.+)/.freeze

    def self.scrub_string(input)
      return input unless input.is_a?(String)
      input
        .gsub(HOME_PATH_RE, "~/")
        .gsub(MEDIA_FILE_RE) { |match| "*#{match[match.rindex(".")..].downcase}" }
        .gsub(SEARCH_TERM_RE, '\1<redacted>')
    end

    def self.scrub_url(input)
      return input unless input.is_a?(String)
      base = input.split("?").first
      base
        .gsub(UUID_RE, "/:id")
        .gsub(NUMERIC_ID_RE, "/:id")
    end

    def self.before_send(event)
      return event unless event
      event[:message] = scrub_string(event[:message]) if event[:message]
      if (req = event[:request])
        req[:url] = scrub_url(req[:url]) if req[:url]
        req[:query_string] = "" if req[:query_string]
      end
      values = event.dig(:exception, :values)
      if values.is_a?(Array)
        values.each do |v|
          v[:value] = scrub_string(v[:value]) if v[:value]
          frames = v.dig(:stacktrace, :frames)
          if frames.is_a?(Array)
            frames.each do |f|
              f[:filename] = scrub_string(f[:filename]) if f[:filename]
              f[:abs_path] = scrub_string(f[:abs_path]) if f[:abs_path]
            end
          end
        end
      end
      if event[:user].is_a?(Hash) && event[:user][:username]
        event[:user].delete(:username)
      end
      event
    end

    def self.before_breadcrumb(crumb)
      return crumb unless crumb
      crumb[:message] = scrub_string(scrub_url(crumb[:message])) if crumb[:message]
      if (data = crumb[:data])
        data[:url] = scrub_url(data[:url]) if data[:url]
        data[:to] = scrub_url(data[:to]) if data[:to]
        data[:from] = scrub_url(data[:from]) if data[:from]
      end
      crumb
    end
  end
end
